lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -275,14 +485,20 @@ def filter_query(
                 if not user_id:
                     raise APIAbort(
                         HTTPStatus.UNAUTHORIZED,
-                        f"Metadata key {k} cannot be used by an unauthenticated client",
+                        f"Metadata key {term.key} cannot be used by an unauthenticated client",
                     )
 
             # NOTE: We don't want to *evaluate* the filter expression here, so
-            # check explicitly for None.
+            # check explicitly for None. I.e., "we have no filter" rather than
+            # "the evaluated result of this filter is falsey".
             if filter is None:
-                expression = aliases[native_key].value[keys].as_string()
-                filter = expression.contains(v) if contains else expression == v
+                expression = (
+                    aliases[native_key]
+                    .value[keys]
+                    .as_string()
+                    .cast(TYPES[vtype].sqltype)
+                )
```

Comment:
You like this as a single expression (as molested by Black) better than, say, a pair of assignments (which might take two lines instead of six)?

Reply:
I'm often amazed at how much I can break SQLAlchemy expressions, but I really wasn't sure of all the intermediaries here and I didn't try. It is fairly ugly, though.